### PR TITLE
Improve position named arguments lint underline and formatting names

### LINKED ISSUE #99907: For position named arguments lint, underline both curly brackets when inserting and don't add $ to formatting names in messages
A comment asked why we only underline the closing curly brace when inserting (https://github.com/rust-lang/rust/pull/99660/files/1b2e05e212f9e2d50fb35817a80136018fbea4ba#r932952838) a named argument. This was originally because that's where it gets inserted -- but on reflection, this is potentially confusing. So underline both.

Also, in the lint message, the formatting argument names are emitted with a dollar sign.
`this formatting argument uses named argument `precision$` by position`
This is also potentially confusing because the actual name does not contain it.

### PR DESCRIPTION
**Pull request by PrestonFrom**
For named arguments used as implicit position arguments, underline both
the opening curly brace and either:
* if there is formatting, the next character (which will either be the
  closing curl brace or the `:` denoting the start of formatting args)
* if there is no formatting, the entire arg span (important if there is
  whitespace like `{  }`)

This should make it more obvious where the named argument should be.

Additionally, in the lint message, emit the formatting argument names
without a dollar sign to avoid potentially confusion.

Fixes #99907

### CODE REVIEW

**Comment by PrestonFrom** [comment]
@lqd This addresses your comment here: https://github.com/rust-lang/rust/pull/99660#discussion_r932952838

Not sure if you want to be the reviewer here or not though?

File: src/test/ui/macros/issue-99265.stderr [hunk 1]
```diff
@@ -190,7 +190,7 @@ LL |         "{}, Hello {1:2$.3$} {4:5$.6$}! {1}",
    |                     - this formatting argument uses named argument `f` by position
 ...
 LL |         f = 0.02f32,
-   |         ^ this named argument is only referred to by position in formatting string
```
Comment (compiler-errors) [inline src/test/ui/macros/issue-99265.stderr:193]: Did you mean to drop "only" here?

File: src/test/ui/macros/issue-99265.stderr [hunk 2]
```diff
@@ -16,9 +16,9 @@ warning: named argument `a` is not used by name
   --> $DIR/issue-99265.rs:9:35
    |
 LL |     println!("{} {} {} {} {}", 0, a=1, b=2, c=3, d=4);
-   |                   -               ^ this named argument is only referred to by position in formatting string
```
Comment (compiler-errors) [inline src/test/ui/macros/issue-99265.stderr:19]: Is the loss of "only" intentional?

File: src/test/ui/macros/issue-99265.stderr (hunk 1 above)
Comment (PrestonFrom) [reply src/test/ui/macros/issue-99265.stderr:193]: With the previous PR, the lint will trigger for every case a named argument is used positionally even if it's used by name elsewhere, so I think "only" is now incorrect. (I missed this in the last change.)

File: src/test/ui/macros/issue-99265.stderr (hunk 2 above)
Comment (compiler-errors) [reply src/test/ui/macros/issue-99265.stderr:19]: duplicate lol, sorry that last note didnt send

File: src/test/ui/macros/issue-99265.stderr (hunk 1 above)
Comment (compiler-errors) [reply src/test/ui/macros/issue-99265.stderr:193]: Ah, ok.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 3]
```diff
@@ -57,26 +58,47 @@ struct PositionalNamedArg {
...
+            (Some(span), Some(span))
         } else if self.ty == PositionalNamedArgType::Arg {
-            // In the case of a named argument whose position is implicit, there will not be a span
-            // to replace. Instead, we insert the name after the `{`, which is the first character
-            // of arg_span.
-            return cx
-                .arg_spans
-                .get(self.cur_piece)
-                .map(|arg_span| arg_span.with_lo(arg_span.lo() + BytePos(1)).shrink_to_lo());
+            // In the case of a named argument whose position is implicit, if the argument *has*
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/format.rs:74]: Thanks for the detailed comment.

File: compiler/rustc_builtin_macros/src/format.rs [hunk 4]
```diff
@@ -57,26 +58,47 @@ struct PositionalNamedArg {
...
+            // formatting, there will not be a span to replace. Instead, we insert the name after
+            // the `{`, which will be the first character of arg_span. If the argument does *not*
+            // have formatting, there may or may not be a span to replace. This is because
+            // whitespace is allowed in arguments without formatting (such as `format!("{  }", 1);`)
+            // but is not allowed in arguments with formatting (an error will be generated in cases
+            // like `format!("{ :1.1}", 1.0f32);`.
+            // For the message span, if there is formatting, we want to use the opening `{` and the
+            // next character, which will the `:` indicating the start of formatting. If there is
+            // not any formatting, we want to underline the entire span.
+            if self.has_formatting {
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/format.rs:84]: Is it possible to factor this into the body of the closure? Both the `if` and `else` have some duplication, so pushing the `if` into the closure might deduplicate some.

Comment (PrestonFrom) [reply compiler/rustc_builtin_macros/src/format.rs]: That's a great idea!

**Comment by PrestonFrom** [comment]
@compiler-errors Thank you for the review!

It looks like #99987 is still in the queue but at the moment I don't think it should have any merge conflicts.

**Comment by compiler-errors** [comment]
Alright, seems like it didn't conflict.

@bors r+

**Pull request merged**